### PR TITLE
Add preparation and followup duration events to calendar

### LINKED ISSUE #3710: Write appointment preparation and follow-up duration as event
**Is your feature request related to a problem? Please describe.**

Appointments let me pick time before and after a potential slot that has to be free for someone to book it. Nevertheless this isn't reflected in the organizer's calendar because we only write the main event with the info visible to the attendee.

**Describe the solution you'd like**

Write additional event(s) into the organizer's calendar if preparation and/or follow-up durations are set. These won't have the attendee added. The attendee must not know about the organizer's preparation time.

**Describe alternatives you've considered**

N/a

**Additional context**

N/a

### PR DESCRIPTION
**Pull request by miaulalala**
FIxes #3710

### CODE REVIEW

File: lib/Service/Appointments/BookingCalendarWriter.php [hunk 1]
```diff
@@ -168,7 +173,54 @@ public function write(AppointmentConfig $config, DateTimeImmutable $start, strin
 		try {
 			$calendar->createFromString($filename . '.ics', $vcalendar->serialize());
 		} catch (CalendarException $e) {
-			throw new RuntimeException('Could not write to calendar', 0, $e);
+			throw new RuntimeException('Could not write event  for appt config id ' . $config->getId(). ' to calendar', 0, $e);
```
Comment (ChristophWurst) [inline lib/Service/Appointments/BookingCalendarWriter.php:176]:
```suggestion
			throw new RuntimeException('Could not write event for appointment config id ' . $config->getId(). ' to calendar', 0, $e);
```

File: lib/Service/Appointments/BookingCalendarWriter.php [hunk 2]
```diff
@@ -168,7 +173,54 @@ public function write(AppointmentConfig $config, DateTimeImmutable $start, strin
...
+		}
+
+		if ($config->getPreparationDuration() !== 0) {
+			$string = $this->l10n->t('Preparation time for %s', [$config->getName()]);
```
Comment (ChristophWurst) [inline lib/Service/Appointments/BookingCalendarWriter.php:180]:
```suggestion
			$string = $this->l10n->t('Prepare for %s', [$config->getName()]);
```

in the hopes of making more text readable on narrow calendar cards

File: lib/Service/Appointments/BookingCalendarWriter.php [hunk 3]
```diff
@@ -168,7 +173,54 @@ public function write(AppointmentConfig $config, DateTimeImmutable $start, strin
...
+					'DTEND' => $start,
+				]
+			]);
+
+			$filename = $this->random->generate(32, ISecureRandom::CHAR_ALPHANUMERIC);
+
+			try {
+				$calendar->createFromString($filename . '.ics', $prepCalendar->serialize());
+			} catch (CalendarException $e) {
+				throw new RuntimeException('Could not write event  for appt config id ' . $config->getId(). ' to calendar', 0, $e);
```
Comment (ChristophWurst) [inline lib/Service/Appointments/BookingCalendarWriter.php:198]:
```suggestion
				throw new RuntimeException('Could not write preparation event for appointment config id ' . $config->getId(). ' to calendar: ' . $e->getMessage(), 0, $e);
```

File: lib/Service/Appointments/BookingCalendarWriter.php [hunk 4]
```diff
@@ -168,7 +173,54 @@ public function write(AppointmentConfig $config, DateTimeImmutable $start, strin
...
+					'DTEND' => $followUpEnd,
+				]
+			]);
+
+			$filename = $this->random->generate(32, ISecureRandom::CHAR_ALPHANUMERIC);
+
+			try {
+				$calendar->createFromString($filename . '.ics', $followUpCalendar->serialize());
+			} catch (CalendarException $e) {
+				throw new RuntimeException('Could not write event  for appt config id ' . $config->getId(). ' to calendar', 0, $e);
```
Comment (ChristophWurst) [inline lib/Service/Appointments/BookingCalendarWriter.php:222]: same as above

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

**Review by ChristophWurst — APPROVED** [review]
:shipit:

**Comment by tcitworld** [comment]
It would be nice to have an `X-NEXTCLOUD-PRE/POST-APPOINTMENT` property for these events so we are able to show them nicely in the calendar, maybe adapt reminders, etc.

Apple has a similar property for travel delay before an event:
```ics
X-APPLE-TRAVEL-DURATION;VALUE=DURATION:PT1H
```

**Comment by ChristophWurst** [comment]
This sounds nice. Shall we put this into a ticket so it doesn't get lost?

https://stackoverflow.com/questions/21497273/subevents-in-icalendar relevant SO topic with the same problem

**Comment by tcitworld** [comment]
https://github.com/nextcloud/calendar/issues/3744

**Pull request merged**